src/common/async/async_cond.h

```diff
@@ -0,0 +1,163 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 IBM
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+/// \file common/async/async_cond.h
+
+#include <cassert>
+#include <concepts>
+#include <mutex>
+#include <utility>
+#include <vector>
+
+#include <boost/asio/any_completion_handler.hpp>
+#include <boost/asio/append.hpp>
+#include <boost/asio/async_result.hpp>
+#include <boost/asio/error.hpp>
+#include <boost/asio/consign.hpp>
+#include <boost/asio/strand.hpp>
+#include <boost/asio/executor_work_guard.hpp>
+
+#include <boost/system/error_code.hpp>
+
+namespace ceph::async {
+/// \brief A non-blocking condition variable
+///
+/// This is effectively a condition variable, but rather than
+/// blocking, the `wait` function takes an Asio completion token and
+/// invokes the associated handler on wakeup.
+template<typename Executor>
+class async_cond {
+  Executor executor;
+  std::mutex m;
+  bool done = false;
+  std::vector<std::pair<
+    boost::asio::any_completion_handler<
+    void(boost::system::error_code)>, std::unique_lock<std::mutex>*>> handlers;
```

Comment:
since this maintains a list of outstanding completion handlers, it should support two-phase shutdown described in [~io_context()](https://www.boost.org/doc/libs/1_82_0/doc/html/boost_asio/reference/io_context/_io_context.html)

consider a coroutine function that puts a `async_cond` variable on its stack, then later waits on it. the resulting completion handler would hold a reference to the coroutine stack preventing its destruction. this is a circular dependency and results in memory leaks

to break that cycle, you can register a `io_context::service` class that gets notified on shutdown. you'd handle that by freeing these handlers with `handlers.clear()`

i added a common/async/service.h for this, with examples uses in common/async/detail/co_throttle_impl.h and spawn_group.h

Reply:
Fixed